File: templates/problems/09_heap/06_design_twitter/go/design_twitter_test.go
package design_twitter

import (
	"reflect"
	"testing"
)

func TestTwitter(t *testing.T) {
	t.Run("basic post and feed", func(t *testing.T) {
		tw := ConstructorTwitter()
		tw.PostTweet(1, 5)
		got := tw.GetNewsFeed(1)
		want := []int{5}
		if !reflect.DeepEqual(got, want) {
			t.Errorf("GetNewsFeed(1) = %v, want %v", got, want)
		}
	})

	t.Run("follow and see followee tweets", func(t *testing.T) {
		tw := ConstructorTwitter()
		tw.PostTweet(1, 5)
		tw.PostTweet(2, 6)
		tw.Follow(1, 2)
		got := tw.GetNewsFeed(1)
		want := []int{6, 5}
		if !reflect.DeepEqual(got, want) {
			t.Errorf("GetNewsFeed(1) = %v, want %v", got, want)
		}
	})

	t.Run("unfollow removes tweets from feed", func(t *testing.T) {
		tw := ConstructorTwitter()
		tw.PostTweet(1, 5)
		tw.PostTweet(2, 6)
		tw.Follow(1, 2)
		tw.Unfollow(1, 2)
		got := tw.GetNewsFeed(1)
		want := []int{5}
		if !reflect.DeepEqual(got, want) {
			t.Errorf("GetNewsFeed(1) = %v, want %v", got, want)
		}
	})

	t.Run("feed limited to 10 most recent", func(t *testing.T) {
		tw := ConstructorTwitter()
		for i := 1; i <= 12; i++ {
			tw.PostTweet(1, i)
		}
		got := tw.GetNewsFeed(1)
		if len(got) != 10 {
			t.Errorf("GetNewsFeed returned %d tweets, want 10", len(got))
		}
		want := []int{12, 11, 10, 9, 8, 7, 6, 5, 4, 3}
		if !reflect.DeepEqual(got, want) {
			t.Errorf("GetNewsFeed(1) = %v, want %v", got, want)
		}
	})

	t.Run("empty feed for new user", func(t *testing.T) {
		tw := ConstructorTwitter()
		got := tw.GetNewsFeed(1)
		if len(got) != 0 {
			t.Errorf("GetNewsFeed(1) = %v, want empty", got)
		}
	})

	t.Run("multiple followees merged feed", func(t *testing.T) {
		tw := ConstructorTwitter()
		tw.PostTweet(2, 10)
		tw.PostTweet(3, 20)
		tw.PostTweet(2, 30)
		tw.Follow(1, 2)
		tw.Follow(1, 3)
		got := tw.GetNewsFeed(1)
		want := []int{30, 20, 10}
		if !reflect.DeepEqual(got, want) {
			t.Errorf("GetNewsFeed(1) = %v, want %v", got, want)
		}
	})
}

File: templates/problems/09_heap/06_design_twitter/typescript/design_twitter.test.ts
import { describe, it, expect } from "vitest";
import { Twitter } from "./design_twitter";

describe("Twitter", () => {
  it("basic post and feed", () => {
    const tw = new Twitter();
    tw.postTweet(1, 5);
    expect(tw.getNewsFeed(1)).toEqual([5]);
  });

  it("follow and see followee tweets", () => {
    const tw = new Twitter();
    tw.postTweet(1, 5);
    tw.postTweet(2, 6);
    tw.follow(1, 2);
    expect(tw.getNewsFeed(1)).toEqual([6, 5]);
  });

  it("unfollow removes tweets from feed", () => {
    const tw = new Twitter();
    tw.postTweet(1, 5);
    tw.postTweet(2, 6);
    tw.follow(1, 2);
    tw.unfollow(1, 2);
    expect(tw.getNewsFeed(1)).toEqual([5]);
  });

  it("feed limited to 10 most recent", () => {
    const tw = new Twitter();
    for (let i = 1; i <= 12; i++) {
      tw.postTweet(1, i);
    }
    expect(tw.getNewsFeed(1)).toEqual([12, 11, 10, 9, 8, 7, 6, 5, 4, 3]);
  });

  it("empty feed for new user", () => {
    const tw = new Twitter();
    expect(tw.getNewsFeed(1)).toEqual([]);
  });

  it("multiple followees merged feed", () => {
    const tw = new Twitter();
    tw.postTweet(2, 10);
    tw.postTweet(3, 20);
    tw.postTweet(2, 30);
    tw.follow(1, 2);
    tw.follow(1, 3);
    expect(tw.getNewsFeed(1)).toEqual([30, 20, 10]);
  });
});
